torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -701,9 +766,26 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
         # Switch to local shard after backward.
         self._use_param_local_shard([param])
 
+        # Prefetch previous layer's full params in backward pass post backward hook,
+        # If next layer's backward computation is done and full params are freed,
+        # no need to prefetch the full params again.
+        # Only prefetch full params if any of the next layer's outputs requires grad
+        if (
+            self.backward_prefetch == BackwardPrefetch_.BACKWARD_POST
+            and self._fsdp_graph_order is not None
+            and self._my_fsdp_idx_in_graph is not None and self._my_fsdp_idx_in_graph > 0
+            and self._fsdp_graph_order[self._my_fsdp_idx_in_graph - 1].training_state != TrainingState_.BACKWARD_POST
+            and self._fsdp_graph_order[self._my_fsdp_idx_in_graph - 1]._need_rebuild_full_params
+        ):
+            self._fsdp_graph_order[self._my_fsdp_idx_in_graph - 1]._rebuild_full_params()  # type: ignore[operator]
+            # Next layer's computation will start right after this all_gather,
+            # Wait for all_gather to finish before computation.
+            torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
```

Comment:
this is because the next layer's pre-backward hook might already be running? If we had guaranteed order such that post backward hook for this layer comes before pre-backward hook for next layer, we wouldn't need this wait?

Reply:
next layer's computation will start immediately, so wait for here. it is like moving up _rebuild_full_params in pre-backward hook here, this move up is to make all gather scheduled before reduce_scatter, so that reduce_scatter is not blocked and can be overlapped with computation